### PR TITLE
Fix a bug in using expected improvement to predict minimization problems

### PR DESCRIPTION
**Pull request by cmjiang**
@joshvfleming @basukinjal @yunboouyang
Tests are done for using EI and CB as prediction transformation to maximize and minimize normal distribution PDF. Other than the title literally, this PR also includes a change which reduce the time of unit tests.

### CODE REVIEW

File: photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/criteria/ExpectedImprovement.scala [hunk 1]
```diff
@@ -33,10 +33,16 @@ import com.linkedin.photon.ml.hyperparameter.estimators.PredictionTransformation
...
 class ExpectedImprovement(evaluator: Evaluator, bestEvaluation: Double) extends PredictionTransformation {
 
+  def isMaxOpt: Boolean = true
+
   private val standardNormal = new Gaussian(0, 1)
 
   /**
    * Applies the expected improvement transformation to the model output
+   *
+   * @param predictiveMeans predictive mean output from the model
```
Comment (basukinjal) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/criteria/ExpectedImprovement.scala]: Thanks for adding these.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearch.scala [hunk 2]
```diff
@@ -170,21 +170,25 @@ class GaussianProcessSearch[T](
   }
 
   /**
-   * Selects the best candidate according to the predicted values, where "best" is determined by the given evaluator
+   * Selects the best candidate according to the predicted values, where "best" is determined by the given transformation
    *
    * @param candidates matrix of candidates
    * @param predictions predicted values for each candidate
+   * @param transformation prediction transformation function
```
Comment (basukinjal) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearch.scala]: Might be better to add a more detailed comment to explain this. As in, in the case of EI, we always search for the max, but for CB, we search for max / min given the optimization problem.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/criteria/ExpectedImprovement.scala [hunk 3]
```diff
@@ -33,10 +33,16 @@ import com.linkedin.photon.ml.hyperparameter.estimators.PredictionTransformation
...
+   * @param predictiveVariances predictive variance output from the model
+   * @return the expected improvement
    */
   def apply(
```
Comment (basukinjal) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/criteria/ExpectedImprovement.scala]: I noticed that the EI does not have a test function which led to the bug before. May be you can add a simple test function for EI and CB to verify the calculations.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearchTest.scala [hunk 4]
```diff
@@ -214,4 +228,15 @@ object GaussianProcessSearchTest {
...
+
+    override protected[ml] val labelAndOffsetAndWeights: RDD[(Long, (Double, Double, Double))] =
+      mock(classOf[RDD[(Long, (Double, Double, Double))]])
+    override val evaluatorType: EvaluatorType = EvaluatorType.RMSE
+    override protected[ml] def evaluateWithScoresAndLabelsAndWeights(
+        scoresAndLabelsAndWeights: RDD[(Long, (Double, Double, Double))]): Double = 0.0
+
+    override def betterThan(score1: Double, score2: Double): Boolean = score1 < score2
+  }
 }
```
Comment (basukinjal) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearchTest.scala]: Add a new line at the end.

Comment (basukinjal) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearchTest.scala]: Add a new line at the end.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearch.scala [hunk 5]
```diff
@@ -170,21 +170,27 @@ class GaussianProcessSearch[T](
...
    */
   protected[search] def selectBestCandidate(
       candidates: DenseMatrix[Double],
-      predictions: DenseVector[Double]): DenseVector[Double] = {
+      predictions: DenseVector[Double],
+      transformation: PredictionTransformation): DenseVector[Double] = {
 
     val init = (candidates(0,::).t, predictions(0))
 
+    val direction = if(transformation.isMaxOpt) 1 else -1
```
Comment (joshvfleming) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/search/GaussianProcessSearch.scala]: Missing space after `if`

**Comment by cmjiang** [comment]
@joshvfleming fixed the space issue. Thank you.

**Pull request merged**